### DISCUSSION TITLE
Is the updater window supposed to show native decorations?

### DISCUSSION THREAD

**Discussion by guihkx** [nuttyartist/notes#471] [General]
Before #469, the updater window looked like this (on Linux + X11):

![image](https://user-images.githubusercontent.com/626206/218450432-5232e226-db7d-47b2-a9f5-30e1e81f2a0c.png)

After #469, we now see native decorations (Linux + X11):

![image](https://user-images.githubusercontent.com/626206/218450692-6b8cd2bb-abf9-4fa7-9912-ba5cf7ff6355.png)

So, which would be the "correct" one? I'm asking because I'm attempting to fix frameless mode on Wayland sessions, and right now the updater window is making Qt draw their own decorations:

![Screenshot from 2023-02-13 11-57-48](https://user-images.githubusercontent.com/626206/218451946-00553cba-f6c7-454d-948e-ce606d8266f2.png)

For comparison, this is how it looks without the Qt decorations on Wayland (i.e. before #469):

![Screenshot from 2023-02-13 12-01-42](https://user-images.githubusercontent.com/626206/218452663-495a4ce3-679f-4ac6-8d04-d9ca26fa7a7d.png)

So, I'm not sure if I should leave it as-is, or if I should revert to the previous look.

Please share your thoughts. Thanks!

**Comment by guihkx** [comment]
I'm also asking because native window decorations are also used in a couple of other places:

- Editor Settings window
- About window

**Comment by zjeffer** [comment]
In my opinion, we should always use native window decorations, everywhere. And not even allow users to disable native window decorations. I know we put a lot of work into making the special macOS-like decorations work with everything but it would fix lots of bugs still left in the code, and would make the overall code structure a lot nicer.

It makes a lot more sense to me that every platform uses their platform's native decorations, instead of having them all have macOS decorations. 

Same thing with frameless windows: I say remove it :)

But I'm probably alone in this..

---

As for the updater window, I would also use native decorations, and fix the styling to match the rest of the app.

**Reply by guihkx** [reply]
While I have no real strong opinions about this, I must say I lean toward the native window decorations team too... :p

Granted, I only started contributing to Notes very recently, but the project itself is fairly mature at this point, and the window management code was written by people much smarter than me, but there's still some annoying bugs lurking around it...

So would it be fair to at least say it's something complicated to maintain? 😄

I don't know, I think we should at least consider using native decorations on Wayland sessions, which currently look kinda bad without the shadows around windows and dialogs...

I think I get where @nuttyartist is coming from, though: Having our app look and behave the same across all OSes, feels awesome. However, in my opinion, this road has major bumps: higher code maintainability, more platform-specific bugs and missing features that we would otherwise have with native window decorations.

I'm obviously not trying to impose anything on @nuttyartist either, I'm just trying to learn your guys' thoughts on this.

**Comment by bjorn** [comment]
As the one who initially patched up Notes to use the native window frame on Linux (ccf71e5760660449e8a4669d154300a6a6355083) and then patched in the option (e013bb14d873e87cbdd9278a4ff18f54300f50f0), I personally prefer the native window frame. However, I also understand the desire to remove it from an aesthetic point of view and the general trend of apps not using the native frame.

I also may prefer the native frame, because the one used by Notes looks like macOS, which just looks out of place on Windows and Linux in my opinion - I don't mind Telegram Desktop rendering its own frame for example, because it actually looks nicer than when it uses the native frame (even when it moves the close button to the other side, but even GNOME's own apps are inconsistent that way for me):

![image](https://user-images.githubusercontent.com/531764/218695161-d265999b-1712-49ce-83d7-97f203a57868.png)

![image](https://user-images.githubusercontent.com/531764/218695221-6c022580-59bf-4fdf-9db9-b1e35acec1a8.png)

But of course, if we do this, and especially if we do this by default, then it should not have annoying bugs. I think we're getting there, but if issues remain that we can't fix on some platforms, then we should consider changing the default.

**Reply by guihkx** [reply]
>which just looks out of place on Windows and Linux in my opinion

I agree. And on Windows specifically they seem a bit more awkward (IMO), because I don't think any Windows user ever had their window controls on the left side... xD

> I think we're getting there

That's also a very good point! I personally need to start using Notes more in frameless mode and with a clean config, just to see what the first-use experience looks like now.

I already have a couple of window management bugs written down in Notes to remind myself :^) I just need to find some time to work on them...

**Comment by nuttyartist** [comment]
First of all, I welcome the current changes to the Updater window, I think it looks great.

Regarding the rest of the discussion, the questions are always about
1. What is most aesthetically pleasing
2. What works best - closer to native in functionality.

And to find the right solution that fits both criteria is very different in each OS. I think a little bit differently from you guys. I think many lay people avoid adopting open-source software because of the lack of care for aesthetics (triumphed by practicality). With Notes, I always I tried to achieve both (a very hard task). But I do think we're getting better in every iteration.

But the purpose of developing good frameless window decorations that feel native on each operating system (yes, it's a challenge, I think currently only macOS is decent) is also to inspire other open source projects (whether future projects by other developers or us) to reuse the same code we worked hard to perfect so that other open source projects can offer high-quality software that competes with commercial ones.

I don't see a problem writing more custom, os dependent code if it's modular enough. See this project that I would like to test: https://github.com/wangwenx190/framelesshelper

**Reply by guihkx** [reply]
You bring in some good points. And to be fair, when I first saw Notes, its looks were mostly what caught my attention. So I'd say aesthetics is an important topic.

>https://github.com/wangwenx190/framelesshelper

Very interesting. And it seems to be well maintained as well...

**Reply by bjorn** [reply]
> Very interesting. And it seems to be well maintained as well...

That may be, though it's pretty clearly written at the top of the readme that both the macOS and Linux implementations are not stable enough, and further a number of known issues are listed. It remains a hard problem, apparently.

**Reply by nuttyartist** [reply]
At least we can use its Windows implementation and either keep our macOS implementation or try to incorporate some of their methods. Anyway, it will require proper testing to see fully how well this fits our app.
